src/com/linkedin/parseq/FusionTask.java

```diff
@@ -248,27 +328,27 @@ protected void propagate(final FusionTraceContext traceContext, final SettablePr
   @Override
   protected Promise<? extends T> run(final Context context) throws Throwable {
     final SettablePromise<T> result = Promises.settable();
-    if (_task == null) {
+    String baseName = getName();
+    if (_asyncTask == null) {
+      /* There is no async predecessor, run propagation immediately */
       FusionTraceContext traceContext = new FusionTraceContext(context,
-          FusionTask.this.getShallowTraceBuilder());
+          FusionTask.this.getShallowTraceBuilder(), baseName);
       propagate(traceContext, result);
     } else {
-      final Task<T> propagationTask = Task.async("fusion", ctx -> {
+      /* There is async predecessor, need to run it first */
+      _shallowTraceBuilder.setName("async fused");
+      _shallowTraceBuilder.setSystemHidden(true);
+      /* PropagationTask will actually run propagation */
+      final Task<T> propagationTask = Task.async(baseName, ctx -> {
         final SettablePromise<T> fusionResult = Promises.settable();
-        FusionTraceContext traceContext = new FusionTraceContext(ctx, FusionTask.this.getShallowTraceBuilder());
+        FusionTraceContext traceContext = new FusionTraceContext(ctx, FusionTask.this.getShallowTraceBuilder(), baseName);
```

Comment:
This line & line 334 are basically the same, right? Or is the ctx different from context?